finat/physically_mapped.py

```diff
@@ -306,3 +340,37 @@ def physical_vertices(self):
 
         :returns: a GEM expression for the physical vertices, shape
                 (gdim, )."""
+
+
+def determinant(A):
+    n = A.shape[0]
+    if n == 0:
+        return 1
+    elif n == 1:
+        return A[0, 0]
+    elif n == 2:
+        return A[0, 0] * A[1, 1] - A[0, 1] * A[1, 0]
+    else:
+        detA = A[0, 0] * determinant(A[1:, 1:])
+        cols = numpy.ones(A.shape[1], dtype=bool)
+        for j in range(1, n):
+            cols[j] = False
+            detA += (-1)**j * A[0, j] * determinant(A[1:][:, cols])
+            cols[j] = True
+        return detA
+
+
+def adjugate(A):
```

Comment:
FWIW, I think we should introduce determinant and related primitives in gem which TSFC can be lower to numerically stable evaluations, especially since people keep on wanting to do 4D stuff.

Reply:
This would go in `/tsfc/gem/gem.py`.  Note: there is already an `Inverse` node in GEM there.  We'd have to track down how it gets lowered.

Reply:
We will address properly on a separate TSFC PR